src/AssetManagement/actions/UniswapV2SwapAction.sol

```diff
@@ -0,0 +1,94 @@
+/**
+ * Created by Arcadia Finance
+ * https://www.arcadia.finance
+ *
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+pragma solidity >=0.8.0 <0.9.0;
+
+import "../ActionBase.sol";
+import "../helpers/UniswapV2Helper.sol";
+import "../utils/ActionAssetData.sol";
+import "../../interfaces/IMainRegistry.sol";
+import "../../interfaces/IVault.sol";
+import "../../interfaces/IERC20.sol";
+
+contract UniswapV2SwapAction is ActionBase, UniswapV2Helper {
+    //Maybe add mainreg address also here
+    constructor(address _router, address _mainreg) ActionBase(_mainreg) UniswapV2Helper(_router) {}
+
+    function executeAction(bytes calldata _actionData) override public {
+        // decode action data
+        (address _vaultAddress, address _caller, bytes memory _actionSpecificData) =
+            abi.decode(_actionData, (address, address, bytes));
+        // preCheck data
+        (actionAssetsData memory _outgoing, actionAssetsData memory _incoming, address[] memory path) = _preCheck(_vaultAddress, _actionSpecificData);
+        // execute Action
+         _execute(_vaultAddress, _outgoing, _incoming, path);
+        // postCheck data
+        _postCheck();
+        // revoke approvals
+        // IVault -> revoke approval for action
+    }
+
+    function _execute(
+        address _vaultAddress,
+        actionAssetsData memory _outgoing,
+        actionAssetsData memory _incoming,
+        address[] memory path
+    ) private {
+                 _uniswapV2Swap(_vaultAddress, _outgoing.assetAmounts[0], _incoming.assetAmounts[0], path);
```

Comment:
will not work without transfers of the assets to address(this). Can this be done better to not have a transfer from the vault to this address, then to the router, then to the pool?

Reply:
right, need to approve action for vault and then transfer assets from vault to the actionHandler. 
The incoming assets however are directly transfered to the vault. 

the only way to avoid that extra transfer would be to have the vault be inherited or composed by all the actions?